### PR TITLE
OSDOCS-7878_4_13:vSphere migration upgrade issues for 4.13

### PR DESCRIPTION
**Pull request by lpettyjo**
Version(s): 4.13 only

Issue: https://issues.redhat.com/browse/OSDOCS-7878

Link to docs preview:
Rel Notes: https://65381--docspreview.netlify.app/openshift-enterprise/latest/release_notes/ocp-4-13-release-notes#ocp-4-13-storage-csi-migration-vsphere-GA
CSI Migration: https://65381--docspreview.netlify.app/openshift-enterprise/latest/storage/container_storage_interface/persistent-storage-csi-migration#persistent-storage-csi-migration-sc-vsphere_persistent-storage-csi-migration
vSphere CSI driver (Important Note change): https://65381--docspreview.netlify.app/openshift-enterprise/latest/storage/container_storage_interface/persistent-storage-csi-vsphere

QE review:
- [x] QE has approved this change.

Additional information:

**PTAL**: @gnufied, @gcharot, @ropatil010,  @sferich888, @xltian, @mwerner2113

### CODE REVIEW

File: modules/persistent-storage-csi-migration-vsphere.adoc [hunk 1]
```diff
@@ -6,34 +6,69 @@
 [id="persistent-storage-csi-migration-sc-vsphere_{context}"]
 = vSphere automatic migration
 
+[IMPORTANT]
+====
+Migration includes significant consequences:
+
+* Enabling migration cannot be undone.
+
+* Migration can take awhile to complete depending on how many nodes are on the cluster.
```
Comment (mburke5678) [inline modules/persistent-storage-csi-migration-vsphere.adoc:15]: Personally, I'm not a fan of _a while_. But, I think it should be two words.
```suggestion
* Migration can take a while (some time??) to complete depending on how many nodes are on the cluster.
```

File: modules/persistent-storage-csi-migration-vsphere.adoc [hunk 2]
```diff
@@ -6,34 +6,69 @@
...
+====
+
 == New installations of {product-title}
 For new installations of {product-title} 4.13, or later, automatic migration is enabled by default.
 
-== Upgrading {product-title}
-When upgrading from {product-title} 4.12, or earlier, to 4.13, automatic CSI migration for vSphere only occurs if you opt in.
+== Upgrading from {product-title} 4.13 to 4.14
+If you are using vSphere in-tree persistent volumes (PVs) and want to upgrade from {product-title} 4.13 to 4.14, update vSphere vCenter and ESXI host to 7.0 Update 3L or 8.0 Update 2, otherwise the {product-title} upgrade is blocked. After updating vSphere, your {product-title} upgrade can occur and automatic migration is enabled by default.
```
Comment (mburke5678) [inline modules/persistent-storage-csi-migration-vsphere.adoc:22]:
```suggestion
If you are using vSphere in-tree persistent volumes (PVs) and want to upgrade from {product-title} 4.13 to 4.14, first (?) update vSphere vCenter and ESXI host to 7.0 Update 3L or 8.0 Update 2, otherwise the {product-title} upgrade is blocked. After updating vSphere, you can upgrade your {product-title} cluster and automatic migration is enabled by default.
```

File: modules/persistent-storage-csi-migration-vsphere.adoc [hunk 3]
```diff
@@ -6,34 +6,69 @@
...
 
 [IMPORTANT]
 ====
-Carefully review the following consequences before opting in to migration:
+If you do *not* update to vSphere 7.0 Update 3L or 8.0 Update 2 and use an administrator acknowledgment to upgrade to {product-title} 4.14, known issues can occur. link:https://access.redhat.com/node/7011683[Before proceeding with the administrator acknowledgement, carefully read this knowledge base article].
+====
 
-* link:https://access.redhat.com/node/7011683[There are known issues that can occur that can cause the migration to fail].
+== Upgrading from {product-title} 4.12 to 4.13
+If you are using vSphere in-tree persistent volumes (PVs) and want to upgrade from {product-title} 4.12 to 4.13, update vSphere vCenter and ESXI host to 7.0 Update 3L or 8.0 Update 2, otherwise the {product-title} upgrade is blocked. After updating vSphere, your {product-title} upgrade can occur and automatic CSI migration for vSphere only occurs if you opt in.
```
Comment (mburke5678) [inline modules/persistent-storage-csi-migration-vsphere.adoc:37]:
```suggestion
If you are using vSphere in-tree persistent volumes (PVs) and want to update from {product-title} 4.12 to 4.13, first update vSphere vCenter and ESXI host to 7.0 Update 3L or 8.0 Update 2, otherwise the {product-title} update is blocked. After updating vSphere, you can update your {product-title} cluster and automatic CSI migration for vSphere only occurs if you opt in.
```

**Comment by mburke5678** [comment]
@lpettyjo A few suggestions. Otherwise LGTM

File: modules/persistent-storage-csi-migration-vsphere.adoc [hunk 4]
```diff
@@ -6,35 +6,69 @@
...
+Alternatively, if you do not want to update vSphere, you can proceed with an {product-title} upgrade by:
 
-* Migration can take awhile to complete depending on how many nodes are on the cluster.
+* Initiating the upgrade from a {product-title} 4.12 Z-stream.
+
+* Performing an administrator acknowledgment by running the following command:
++
+[source, cli]
+----
+oc -n openshift-config patch cm admin-acks --patch '{"data":{"ack-4.12-kube-1.26-api-removals-in-4.14":"true"}}' --type=merge
```
Comment (gcharot) [inline modules/persistent-storage-csi-migration-vsphere.adoc:47]: wrong key
cc @gnufied

Comment (gnufied) [reply modules/persistent-storage-csi-migration-vsphere.adoc:47]: Why are we documenting 4.12 upgrades here though? Was it for EUS to EUS scenario?

Comment (gnufied) [reply modules/persistent-storage-csi-migration-vsphere.adoc:47]: Also correct key is - `ack-4.12-kube-126-vsphere-migration-in-4.14`

File: modules/persistent-storage-csi-migration-vsphere.adoc [hunk 5]
```diff
@@ -6,35 +6,61 @@
...
+====
+
 == New installations of {product-title}
 For new installations of {product-title} 4.13, or later, automatic migration is enabled by default.
 
-[id="upgrading-openshift_{context}"]
-== Upgrading {product-title}
-When upgrading from {product-title} 4.12, or earlier, to 4.13, automatic CSI migration for vSphere only occurs if you opt in.
+== Updating from {product-title} 4.13 to 4.14
+If you are using vSphere in-tree persistent volumes (PVs) and want to update from {product-title} 4.13 to 4.14, first update vSphere vCenter and ESXI host to 7.0 Update 3L or 8.0 Update 2, otherwise the {product-title} update is blocked. After updating vSphere, your {product-title} update can occur and automatic migration is enabled by default.
```
Comment (gcharot) [inline modules/persistent-storage-csi-migration-vsphere.adoc:22]: unless CSI migration is already enabled
cc @gnufied

File: modules/persistent-storage-csi-migration-vsphere.adoc [hunk 6]
```diff
@@ -6,35 +6,61 @@
...
++
+[source, cli]
+----
+oc -n openshift-config patch cm admin-acks --patch '{"data":{"ack-4.12-kube-126-vsphere-migration-in-4.14":"true"}}' --type=merge
+----
 
-* Migration is a disruptive process. The Machine Config Operator (MCO) has to update kubelet, which means degrading each node as it rolls out the new MachineConfig to each machine.
+[IMPORTANT]
+====
+If you update to {product-title} 4.13 without updating vSphere, and then opt in to migration, known issues can occur. Review the consequences listed above for migration before opting in and link:https://access.redhat.com/node/7011683[carefully read this knowledge base article].
```
Comment (gcharot) [inline modules/persistent-storage-csi-migration-vsphere.adoc:52]: add link to the opt in doc ? https://docs.openshift.com/container-platform/4.13/storage/container_storage_interface/persistent-storage-csi-migration.html#upgrading-openshift_persistent-storage-csi-migration

File: modules/persistent-storage-csi-migration-vsphere.adoc [hunk 7]
```diff
@@ -6,35 +6,69 @@
...
-== Upgrading {product-title}
-When upgrading from {product-title} 4.12, or earlier, to 4.13, automatic CSI migration for vSphere only occurs if you opt in.
+== Upgrading from {product-title} 4.13 to 4.14
+If you are using vSphere in-tree persistent volumes (PVs) and want to upgrade from {product-title} 4.13 to 4.14, update vSphere vCenter and ESXI host to 7.0 Update 3L or 8.0 Update 2, otherwise the {product-title} upgrade is blocked. After updating vSphere, your {product-title} upgrade can occur and automatic migration is enabled by default.
+
+Alternatively, if you do not want to update vSphere, you can proceed with an {product-title} upgrade by performing an administrator acknowledgment:
+
+[source, cli]
+----
+oc -n openshift-config patch cm admin-acks --patch '{"data":{"ack-4.13-kube-1.27-api-removals-in-4.14":"true"}}' --type=merge
```
[the remark this answers is not in the document]
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-vsphere.adoc]: Corrected.

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 4 above)
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-vsphere.adoc]: Strictly speaking we don't have to include this content here. I think that we were doing this to ensure absolute coverage by including it in the 4.12 and 4.13 docs.

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 5 above)
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-vsphere.adoc:22]: Correct me if I'm wrong, but if you have enabled migration, then you won't have in-tree PVs because they have been migrated to CSI, so the first phrase, I think, covers this. @gnufied @gcharot

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 6 above)
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-vsphere.adoc]: Done.

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 5 above)
Comment (gcharot) [reply modules/persistent-storage-csi-migration-vsphere.adoc]: The in-tree PVs still appear at in-tree but yes fair let's not complexify the wording.

File: modules/persistent-storage-csi-migration-vsphere-to-4-13.adoc [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+If you are using vSphere in-tree persistent volumes (PVs) and want to update from {product-title} 4.12 to 4.13, update vSphere vCenter and ESXI host to 7.0 Update 3L or 8.0 Update 2, otherwise the {product-title} update is blocked. After updating vSphere, your {product-title} update can occur and automatic CSI migration for vSphere only occurs if you opt in.
+
+Alternatively, if you do not want to update vSphere, you can proceed with an {product-title} update by running the following command to perform an administrator acknowledgment:
+
+[source, cli]
+----
+oc -n openshift-config patch cm admin-acks --patch '{"data":{"ack-4.12-kube-126-vsphere-migration-in-4.14":"true"}}' --type=merge
+----
+
```
Comment (gnufied) [inline modules/persistent-storage-csi-migration-vsphere-to-4-13.adoc:17]: I would like us to incorporate same wording as - https://github.com/openshift/openshift-docs/pull/65546#discussion_r1370294809 

because tbh, we want folks to upgrade to 4.13 and not discourage it.

Comment (gcharot) [reply modules/persistent-storage-csi-migration-vsphere-to-4-13.adoc:17]: Agreed maybe rephrase the important square below to make it more positive.

Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-vsphere-to-4-13.adoc:17]: Done.

File: modules/persistent-storage-csi-migration-vsphere-to-4-14.adoc [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+Alternatively, if you do not want to update vSphere, you can proceed with an {product-title} update by performing an administrator acknowledgment:
+
+[source, cli]
+----
+oc -n openshift-config patch cm admin-acks --patch '{"data":{"ack-4.13-kube-127-vsphere-migration-in-4.14":"true"}}' --type=merge
+----
+
+[IMPORTANT]
+====
+If you do *not* update to vSphere 7.0 Update 3L or 8.0 Update 2 and use an administrator acknowledgment to update to {product-title} 4.14, known issues can occur. link:https://access.redhat.com/node/7011683[Before proceeding with the administrator acknowledgement, carefully read this knowledge base article].
```
[the remark this answers is not in the document]
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-vsphere-to-4-14.adoc:20]: Done.

**Comment by gnufied** [comment]
lgtm

**Comment by ropatil010** [comment]
LGTM, Thanks.!
cc @duanwei33

**Comment by gcharot** [comment]
LGTM

**Comment by mwerner2113** [comment]
LGTM

**Comment by kalexand-rh** [comment]
@xltian, will you please ack this PR that's out for change management?

**Comment by duanwei33** [comment]
The scenario has been verified by QE.

**Comment by xltian** [comment]
Thanks, Wei, LGTM too.
